File: S-SingleResponsibility/Good_UserService.cs
// ✅ GOOD — Follows Single Responsibility Principle
// Each class now has exactly ONE reason to change.

// Responsibility 1: Only handles database operations
public class UserRepository
{
    public void Save(string username)
        => Console.WriteLine($"Saving {username} to database...");
}

// Responsibility 2: Only handles email sending
public class EmailService
{
    public void SendWelcomeEmail(string email)
        => Console.WriteLine($"Sending welcome email to {email}...");
}

// Responsibility 3: Only handles logging
public class Logger
{
    public void Log(string message)
        => Console.WriteLine($"LOG: {message}");
}

// Orchestrator: only coordinates the above services
public class UserService
{
    private readonly UserRepository _repo = new();
    private readonly EmailService _email = new();
    private readonly Logger _logger = new();

    public void RegisterUser(string username, string email)
    {
        _repo.Save(username);
        _email.SendWelcomeEmail(email);
        _logger.Log($"User {username} registered.");
    }
}

// 💡 Now if the email provider changes, only EmailService changes.
// The registration logic and DB logic are completely untouched.

File: S-SingleResponsibility/Good_UserService.cpp
// ✅ GOOD — Follows Single Responsibility Principle
// Each class now has exactly ONE reason to change.

#include <iostream>
#include <string>
#include <memory>

// Responsibility 1: Only handles database operations
class UserRepository {
public:
    void Save(const std::string& username) {
        std::cout << "Saving " << username << " to database...\n";
    }
};

// Responsibility 2: Only handles email sending
class EmailService {
public:
    void SendWelcomeEmail(const std::string& email) {
        std::cout << "Sending welcome email to " << email << "...\n";
    }
};

// Responsibility 3: Only handles logging
class Logger {
public:
    void Log(const std::string& message) {
        std::cout << "LOG: " << message << "\n";
    }
};

// Orchestrator: only coordinates the above services
class UserService {
    UserRepository _repo;
    EmailService   _email;
    Logger         _logger;

public:
    void RegisterUser(const std::string& username, const std::string& email) {
        _repo.Save(username);
        _email.SendWelcomeEmail(email);
        _logger.Log("User " + username + " registered.");
    }
};

// 💡 Now if the email provider changes, only EmailService changes.
// The registration logic and DB logic are completely untouched.
